File: eunji/0x07_deque/BOJ_5430.py
from collections import deque
import sys

def print_deq(D, flag):
    if (flag == -1): print("error")
    elif (len(D) == 0): print("[]")
    elif (flag == 0): 
        print('[', end='')
        for i in range(len(D)-1):
            print(str(D[i])+',', end='')
        print(str(D[len(D)-1])+']')
    else:
        print('[',end='')
        for i in range(len(D)-1,0,-1):
            print(str(D[i]), end=',')
        print(str(D[0])+']')

def make_deq(D, size, ns):
    if size == 0:
        return
    ns = ns[1:-2].split(',') 
    for i in ns:
        if i == ',': continue
        D.append(int(i))   
        
testcase = int(sys.stdin.readline())
for i in range(testcase):
    reverse_flag = 0
    func = sys.stdin.readline()
    size = int(sys.stdin.readline())
    ns = sys.stdin.readline()
    D = deque([])
    make_deq(D, size, ns)
    for f in func:
        if (f == 'R'):
            if (reverse_flag == 0): reverse_flag = 1
            else: reverse_flag = 0
        elif (f == 'D'):
            if (len(D) == 0):
                reverse_flag = -1
                break
            elif (reverse_flag == 0): D.popleft()
            elif (reverse_flag == 1): D.pop()
            
    print_deq(D, reverse_flag)

File: eunji/0x07_deque/BOJ_5430.cpp
#include <iostream>
#include <deque>
#include <sstream>
using namespace std;

void print_deq(deque<int> &D, int flag){
    if (flag == -1) cout << "error" << '\n';
    else if (D.size() == 0) cout << "[]" << '\n';
    else if (flag == 0){
        cout << '[';
        for (int i=0; i<D.size()-1; i++){
            cout << D[i] << ',';
        }
        cout << D[D.size()-1] << ']' << '\n';
    }
    else{
        cout << '[';
        for (int i=D.size()-1; i>0; i--){
            cout << D[i] << ',';
        }
        cout << D[0] << ']' << '\n';

    }
}
void make_deq(deque<int> &D, int size, string ns){
    size_t pos = ns.find(']');
    ns = ns.substr(1, pos-1);
    stringstream ss(ns);
    string temp;
    while (getline(ss, temp, ',')){
        D.push_back(stoi(temp));
    }
}

int main(void){
    ios::sync_with_stdio(0);
    cin.tie(0);
    int testcase;
    cin >> testcase;
    for (int i=0; i<testcase; i++){
        int reverse_flag = 0;
        string func; 
        cin >> func;
        int size;   
        cin >> size;
        string ns;
        cin >> ns;
        deque<int> D;
        make_deq(D, size, ns);
        for (auto f: func){
            if (f == 'R') {
                if (reverse_flag == 0) reverse_flag = 1;
                else reverse_flag = 0;
            }
            else if (f == 'D'){
                if (D.size() == 0) {
                    reverse_flag = -1; 
                    break;
                }
                else if (reverse_flag == 0) D.pop_front();
                else if (reverse_flag == 1) D.pop_back();
            }
        }
        print_deq(D, reverse_flag);   
    }
}